File: src/main/java/guru/xgm/language/angelscript/lang/folding/AngelScriptCustomFoldingBuilder.java
package guru.xgm.language.angelscript.lang.folding;

import com.intellij.lang.ASTNode;
import com.intellij.lang.folding.CustomFoldingBuilder;
import com.intellij.lang.folding.FoldingDescriptor;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiComment;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiTreeUtil;
import guru.xgm.language.angelscript.extapi.psi.AngelScriptPsiFileBase;
import guru.xgm.language.angelscript.lang.AngelScriptParserDefinition;
import guru.xgm.language.angelscript.psi.AngelScriptEnums;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.List;

import static guru.xgm.language.angelscript.psi.AngelScriptTypes.ENUMS;

final class AngelScriptCustomFoldingBuilder extends CustomFoldingBuilder implements DumbAware {

    @Override
    protected boolean isCustomFoldingRoot(@NotNull final ASTNode node) {
        final IElementType type = node.getElementType();
        return type == AngelScriptParserDefinition.ANGELSCRIPT_FILE || type == ENUMS;
    }

    @Override
    protected void buildLanguageFoldRegions(@NotNull List<FoldingDescriptor> descriptors, @NotNull PsiElement root, @NotNull Document document, boolean quick) {
        if (!(root instanceof AngelScriptPsiFileBase)) return;

        final Collection<PsiElement> psiElements = PsiTreeUtil.findChildrenOfAnyType(
                root,
                PsiComment.class,
                AngelScriptEnums.class
        );

        for (PsiElement element : psiElements) {
            if (element instanceof AngelScriptEnums enums) {
                final var id = enums.getId();

                final var s = id.getNode();
                final var e = enums.getNode();

                descriptors.add(new FoldingDescriptor(element, TextRange.create(s.getStartOffset() + s.getTextLength(), e.getStartOffset() + e.getTextLength())));
            }
        }
    }

    @Override
    protected String getLanguagePlaceholderText(@NotNull ASTNode node, @NotNull TextRange range) {
        final IElementType type = node.getElementType();

        if (type == ENUMS) {
            final var psi = node.getPsi(AngelScriptEnums.class);
            final var block = psi.getEnumStatBlock();
            return block == null ? "..." : sizeable(block.getEnumItemList());
        }
        return null;
    }

    private String sizeable(List<?> list) {
        final int size = list.size();
        return size == 0 ? " ... " : " (" + size + ") ";
    }

    @Override
    protected boolean isRegionCollapsedByDefault(@NotNull ASTNode node) {
        final IElementType type = node.getElementType();
        final var settings = AngelScriptCodeFoldingSettings.getInstance();

        if (type == ENUMS) return settings.isFoldEnum();
        return false;
    }

    private void foldSimple(PsiElement element, IElementType start, IElementType end, List<FoldingDescriptor> descriptors) {
        final ASTNode s = element.getNode().findChildByType(start);
        final ASTNode e = element.getNode().findChildByType(end, s);

        if (s == null || e == null) return;

        descriptors.add(new FoldingDescriptor(element, TextRange.create(s.getStartOffset() + s.getTextLength(), e.getStartOffset())));
    }

}

File: src/main/kotlin/raft/war/language/angelscript/lang/folding/AngelScriptCustomFoldingBuilder.kt
package raft.war.language.angelscript.lang.folding

import com.intellij.lang.ASTNode
import com.intellij.lang.folding.CustomFoldingBuilder
import com.intellij.lang.folding.FoldingDescriptor
import com.intellij.openapi.editor.Document
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiComment
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.util.elementType
import com.intellij.psi.util.startOffset
import raft.war.ide.IdeMultiHostInjector.Companion.INJECT_JASS
import raft.war.language.angelscript.psi.file.AngelScriptFile
import raft.war.language.angelscript.lang.ANGELSCRIPT_FILE
import raft.war.language.angelscript.lang.folding.AngelScriptCodeFoldingSettings.Companion.instance
import raft.war.language.angelscript.psi.AngelScriptTypes.ENUMS
import raft.war.language.angelscript.psi.AngelScriptTypes.LINE_COMMENT

internal class AngelScriptCustomFoldingBuilder : CustomFoldingBuilder(), DumbAware {
    override fun isCustomFoldingRoot(node: ASTNode): Boolean {
        val type = node.elementType
        return type === ANGELSCRIPT_FILE || type === ENUMS
    }

    override fun buildLanguageFoldRegions(
        descriptors: MutableList<FoldingDescriptor>,
        root: PsiElement,
        document: Document,
        quick: Boolean
    ) {
        if (root !is AngelScriptFile) return

        val psiElements = PsiTreeUtil.findChildrenOfAnyType(
            root,
            PsiComment::class.java,
            raft.war.language.angelscript.psi.AngelScriptEnums::class.java
        )

        for (element in psiElements) {
            val type = element.elementType

            if (type == LINE_COMMENT && element.text.startsWith(INJECT_JASS)) {
                descriptors.add(
                    FoldingDescriptor(
                        element,
                        TextRange.create(element.startOffset, element.startOffset + 3)
                    )
                )
            }

            if (element is raft.war.language.angelscript.psi.AngelScriptEnums) {
                val id: PsiElement = element.id

                val s = id.node
                val e = element.getNode()

                descriptors.add(
                    FoldingDescriptor(
                        element,
                        TextRange.create(s.startOffset + s.textLength, e.startOffset + e.textLength)
                    )
                )
            }
        }
    }

    override fun getLanguagePlaceholderText(node: ASTNode, range: TextRange): String? {
        val type = node.elementType

        if (type === LINE_COMMENT) {
            return ""
        }

        if (type === ENUMS) {
            val psi = node.getPsi(raft.war.language.angelscript.psi.AngelScriptEnums::class.java)
            val block = psi.enumStatBlock
            return if (block == null) "..." else sizeable(block.enumItemList)
        }
        return null
    }

    private fun sizeable(list: List<*>): String {
        val size = list.size
        return if (size == 0) " ... " else " ($size) "
    }

    override fun isRegionCollapsedByDefault(node: ASTNode): Boolean {
        val type = node.elementType
        val settings = instance

        return when {
            type === ENUMS -> settings.isFoldEnum
            type === LINE_COMMENT -> true
            else -> false
        }
    }
}
